### PR TITLE
Faster consistency checker

### PR DESCRIPTION
**Pull request by tinwelint**
Basically by reducing random access using an effient cache, checking only
parts of records through multiple passes. Also parallelizes some store
checks, like indexes.

### CODE REVIEW

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/RelationshipRecordCheck.java [hunk 1]
```diff
@@ -19,32 +19,77 @@
...
+                RelationshipField.SOURCE_PREV, RelationshipField.TARGET_NEXT, RelationshipField.TARGET_PREV,
+                RelationshipField.CACHE_VALUES );
+    }
+
+    public static RelationshipRecordCheck RelationshipRecordCheckTargetChain( boolean firstProperty )
+    {
+        return new RelationshipRecordCheck( firstProperty, RelationshipField.TARGET_NEXT, RelationshipField.TARGET_PREV );
+    }
+
+    private enum RelationshipTypeField
```
Comment (davidegrohmann) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/RelationshipRecordCheck.java:84]: can we extract this enum?

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/RelationshipRecordCheck.java [hunk 2]
```diff
@@ -274,23 +609,43 @@ public void checkReference( RelationshipRecord record, RelationshipRecord referr
             }
             else
             {
+                CacheAccess.Client cacheAccess = records.cacheAccess().client();
                 if ( other( field, referred ) != record.getId() )
                 {
-                    noBackReference( engine.report(), referred );
+                    if ( !referred.isReal() )//&& isSkippable(record, cacheFields))
```
Comment (davidegrohmann) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/RelationshipRecordCheck.java:615]: please remove commented code

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheAccess.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+ *
+ * The cache is a compact representation of records, tied to an id, for example nodeId. There can be multiple
+ * cached values per id, selected by {@code slot}.
+ */
+public interface CacheAccess
+{
+    /**
+     * Client per thread for accessing cache and counts for statistics
+     */
+    interface Client
```
Comment (davidegrohmann) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheAccess.java:52]: move this out in its own file?

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheProcessor.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+import org.neo4j.kernel.impl.store.record.PropertyKeyTokenRecord;
+import org.neo4j.kernel.impl.store.record.PropertyRecord;
+import org.neo4j.kernel.impl.store.record.RelationshipGroupRecord;
+import org.neo4j.kernel.impl.store.record.RelationshipRecord;
+import org.neo4j.kernel.impl.store.record.RelationshipTypeTokenRecord;
+
+/**
+ * A work-around for embedding a {@link CacheAction} inside a {@link StoreProcessor}.
+ */
+public class CacheProcessor extends StoreProcessor
```
Comment (davidegrohmann) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheProcessor.java:49]: This smells of using the wrong abstraction.

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/PropertyAndNodeIndexedCheck.java [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+            }
+            else
+            {
+                if ( field.otherReference( referred ) != record.getId() )
+                {
+                    field.noBackReference( engine.report(), referred );
+                }
+            }
+        }
+    }*/
```
Comment (davidegrohmann) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/PropertyAndNodeIndexedCheck.java:225]: please remove commented code

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/RecordScanner.java [hunk 6]
```diff
@@ -66,15 +98,66 @@ public void run()
...
+        {
+            super( recordsQ );
+            this.processor = processor;
+        }
+
+        @Override
+        protected void process( RECORD record )
+        {
+            processor.process( record );
+        }
```
Comment (davidegrohmann) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/RecordScanner.java:161]: Can we split this in 3 parts? 1) SequentialRecordScanner, 2) a ParallelRecordScanner and 3) a CompositeRecordScanner that picks the first or the second based on the stage?

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/StoreProcessor.java [hunk 7]
```diff
@@ -145,4 +205,80 @@ public void processSchema( RecordStore<DynamicRecord> store, DynamicRecord schem
...
+                break;
+            }
+            progressListener.set( entryCount++ );
+        }
+        progressListener.done();
+        for ( Worker<R> worker : workers )
+        {
+            worker.done();
+        }
+        workers.awaitAndThrowOnError( RuntimeException.class );
```
Comment (davidegrohmann) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/StoreProcessor.java:265]: perhaps we can extract this in its own class? so we can test it

**Comment by tinwelint** [comment]
@davidegrohmann I'll close this and reissue for 2.3, but will deal with your comments there instead. Please add new comments to the new PR, when it hits.

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/RelationshipRecordCheck.java (hunk 1 above)
Comment (tinwelint) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/RelationshipRecordCheck.java:84]: It seems to be a common pattern here that *RecordCheck classes have their enum fields as internal classes... all except NodeField which is an enum for RelationshipRecordCheck. It's not strictly necessary, is it? We unit test on RecordCheck level anyway.

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheAccess.java (hunk 3 above)
Comment (tinwelint) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheAccess.java:52]: I think it makes sense to have it in `CacheAccess` because as it looks then `Client` doesn't make sense without `CacheAccess`.

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheProcessor.java (hunk 4 above)
Comment (tinwelint) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheProcessor.java:49]: I agree, we should try to change it

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/RecordScanner.java (hunk 6 above)
Comment (tinwelint) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/RecordScanner.java:161]: That makes sense, will do

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheAccess.java (hunk 3 above)
Comment (davidegrohmann) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheAccess.java:52]: can we use packages for grouping instead subclassing?

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/RelationshipRecordCheck.java (hunk 1 above)
Comment (davidegrohmann) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/RelationshipRecordCheck.java:84]: Not is is not strictly needed, but it would be nice to reduce complexity. Having classes of 677 lines it is something we should try to avoid.

This is again using subclassing instead of packages for grouping things together.

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/StoreProcessor.java (hunk 7 above)
Comment (tinwelint) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/StoreProcessor.java:265]: Simplified it quite a bit by removing the `Distributor` which wasn't really needed.

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/RelationshipRecordCheck.java (hunk 1 above)
Comment (tinwelint) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/RelationshipRecordCheck.java:84]: Yeah, or internal classes you mean. We could do this, but let it be part of a separate commit or PR altogether. Otherwise it's so much harder to reason about the changes in this PR.

File: community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheAccess.java (hunk 3 above)
Comment (tinwelint) [inline community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheAccess.java:52]: Internal classes/interfaces has its use, just like packages have. Look at what we usually do with `Adapter`, `Delegator`, `Monitor` or `Configuration`... all internal interfaces that makes those small interfaces easier to understand when they're internal, also reads better when spelling out, f.ex. `CacheAccess.Client`.
